### PR TITLE
Add `plot_features_relevance`

### LINKED ISSUE #564: Feature relevance visualisation
### 🚀 Feature Request

Create method to visualise feature relevance 

### Motivation

Tool for feature analysis

### Proposal

Create method
```
def plot_feature_relevance(
    ts: TSDataset,  
    relevance_table: RelevanceTable,
    normalized: bool = False,
    relevance_aggregation_mode: Union[str, Literal["per-segment"]] = AggregationMode.mean,  
    top_k: Optional[int] = None,
    segments: Optional[List[str]] = None,
    columns_num: int = 2,
    figsize: Tuple[int, int] = (10, 5)
)
```
Where
relevance_table - method to evaluate the feature relevance
normalized - whether obtained relevances should be normalized
relevance_aggregation_mode - aggregation strategy for obtained feature relevance table, is 
top_k - number of best features to plot, if None plot all the features

- Method should use relevance_table to evaluate feature relevance
- Plot the the sorted feature relevance 
- If `relevance_aggregation_mode="per-segment"` plot feature relevance for each segment, else aggregate feature relevance using modes from `AggregationMode`(see mrmr)
- Sort features depending on `RelevanceTable.greater_is_better`

### Test cases

_No response_

### Alternatives

_No response_

### Additional context

__results___38_0 (1)

### Checklist

- [ ] I discussed this issue with ETNA Team

### PR DESCRIPTION
**Pull request by Mr-Geekman**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Proposed Changes
Look #564.

## Related Issue
#564.

## Closing issues
Closes #564.

### CODE REVIEW

**Comment by Mr-Geekman** [comment]
Script for demonstration:
```python
from typing import Tuple

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from etna.analysis import StatisticsRelevanceTable
from etna.analysis import plot_feature_relevance
from etna.datasets import TSDataset


def simple_df_relevance() -> Tuple[pd.DataFrame, pd.DataFrame]:
    rng = np.random.default_rng(42)
    timestamp = pd.date_range("2021-01-01", "2021-02-01")

    df_1 = pd.DataFrame({"timestamp": timestamp, "target": np.arange(32), "segment": "1"})
    df_2 = pd.DataFrame({"timestamp": timestamp[5:], "target": np.arange(5, 32), "segment": "2"})
    df = pd.concat([df_1, df_2], ignore_index=True)
    df = TSDataset.to_dataset(df)

    timestamp = pd.date_range("2020-12-01", "2021-02-11")

    df_1 = pd.DataFrame(
        {
            "timestamp": timestamp,
            "regressor_1": np.arange(len(timestamp)),
            "regressor_2": np.zeros(len(timestamp)),
            "regressor_3": rng.normal(loc=0, scale=1.0, size=len(timestamp)),
            "regressor_4": rng.uniform(low=0, high=10.0, size=len(timestamp)),
            "regressor_5": rng.exponential(scale=1.0, size=len(timestamp)),
            "cat_feature": "hello",
            "segment": "1",
        }
    )
    df_2 = pd.DataFrame(
        {
            "timestamp": timestamp[5:],
            "regressor_1": np.sin(-np.arange(len(timestamp) - 5)),
            "regressor_2": np.log(np.arange(1, len(timestamp) - 4)),
            "regressor_3": rng.normal(loc=0, scale=2.0, size=len(timestamp) - 5),
            "regressor_4": rng.uniform(low=0, high=5.0, size=len(timestamp) - 5),
            "regressor_5": rng.exponential(scale=0.1, size=len(timestamp) - 5),
            "cat_feature": "bye",
            "segment": "2",
        }
    )
    df_exog = pd.concat([df_1, df_2], ignore_index=True)
    df_exog = TSDataset.to_dataset(df_exog)

    return df, df_exog


df, df_exog = simple_df_relevance()
ts = TSDataset(df=df, df_exog=df_exog, known_future="all", freq="D")
relevance_table = StatisticsRelevanceTable()

plot_feature_relevance(
    ts=ts,
    relevance_table=relevance_table,
    normalized=False,
    relevance_aggregation_mode="per-segment",
    top_k=None,
    segments=None,
    columns_num=2,
    figsize=(10, 5),
)

plt.savefig("per-segment.png")

plot_feature_relevance(
    ts=ts,
    relevance_table=relevance_table,
    normalized=False,
    relevance_aggregation_mode="mean",
    top_k=None,
    segments=None,
    columns_num=2,
    figsize=(10, 5),
)

plt.savefig("mean.png")
```

mean:
mean

per-segment:
per-segment

File: etna/analysis/plotters.py [hunk 1]
```diff
@@ -729,3 +732,64 @@ def plot_residuals(
...
+    Parameters
+    ----------
+    ts:
+        TSDataset with timeseries data
+    relevance_table:
+        method to evaluate the feature relevance
+    normalized:
+        whether obtained relevances should be normalized to sum up to 1
+    relevance_aggregation_mode:
+        aggregation strategy for obtained feature relevance table
```
Comment (alex-hse-repository) [inline etna/analysis/plotters.py:759]: May be it is better to list the possible variants of aggregation

File: etna/analysis/plotters.py [hunk 2]
```diff
@@ -729,3 +732,64 @@ def plot_residuals(
...
+    columns_num:
+        if `relevance_aggregation_mode="per-segment"` number of columns in subplots, otherwise the value is ignored
+    figsize:
+        size of the figure per subplot with one segment in inches
+    """
+    if not segments:
+        segments = sorted(ts.segments)
+
+    is_ascending = not relevance_table.greater_is_better
+    relevance_df = relevance_table(df=ts[:, :, "target"], df_exog=ts.df_exog).loc[segments]
```
Comment (alex-hse-repository) [inline etna/analysis/plotters.py:773]: df_exog does not contain all the features in the dataset, you need to use `ts[:,:,features]` here, where `features = set(ts.columns.get_level_values("feature")) - set(["target"])`

File: etna/analysis/plotters.py [hunk 3]
```diff
@@ -729,3 +732,64 @@ def plot_residuals(
...
+def plot_feature_relevance(
+    ts: "TSDataset",
+    relevance_table: RelevanceTable,
+    normalized: bool = False,
+    relevance_aggregation_mode: Union[str, Literal["per-segment"]] = AggregationMode.mean,
+    top_k: Optional[int] = None,
+    segments: Optional[List[str]] = None,
+    columns_num: int = 2,
+    figsize: Tuple[int, int] = (10, 5),
+):
```
Comment (alex-hse-repository) [inline etna/analysis/plotters.py:746]: Add somewhere `**relevance_params` because `ModelRelevanceTable` requires the model to be specified

File: etna/analysis/plotters.py [hunk 4]
```diff
@@ -729,3 +732,64 @@ def plot_residuals(
...
+    """
+    Plot relevance of the features.
```
Comment (alex-hse-repository) [inline etna/analysis/plotters.py:748]: May you clarify that the features on the plot are order in descending order of their importance

File: etna/analysis/plotters.py [hunk 5]
```diff
@@ -729,3 +732,64 @@ def plot_residuals(
...
+                relevance = relevance / relevance.sum()
+            sns.barplot(x=relevance.values, y=relevance.index, orient="h", ax=ax[i])
+            ax[i].set_title(f"Feature relevance: {segment}")
+
+    else:
+        relevance_aggregation_fn = AGGREGATION_FN[AggregationMode(relevance_aggregation_mode)]
+        relevance = relevance_df.apply(lambda x: relevance_aggregation_fn(x[~x.isna()]))  # type: ignore
+        relevance = relevance.sort_values(ascending=is_ascending)
+        # if top_k == None, all the values are selected
+        relevance = relevance.dropna()[:top_k]
```
Comment (alex-hse-repository) [inline etna/analysis/plotters.py:790]: May be let's throw warning that we are not able to calculate the relevance of some features

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]
May you also try to test this method in the different combinations of the parameters(different relevance tables, normalization, top_k, aggregation mod)

**Comment by Mr-Geekman** [comment]
Скрипт:
```python
from typing import Tuple

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestRegressor

from etna.analysis import ModelRelevanceTable
from etna.analysis import StatisticsRelevanceTable
from etna.analysis import plot_feature_relevance
from etna.datasets import TSDataset
from etna.transforms import DateFlagsTransform
from etna.transforms import FilterFeaturesTransform


def simple_df_relevance() -> Tuple[pd.DataFrame, pd.DataFrame]:
    rng = np.random.default_rng(42)
    timestamp = pd.date_range("2021-01-01", "2021-02-01")

    df_1 = pd.DataFrame({"timestamp": timestamp, "target": np.arange(32), "segment": "1"})
    df_2 = pd.DataFrame({"timestamp": timestamp[5:], "target": np.arange(5, 32), "segment": "2"})
    df = pd.concat([df_1, df_2], ignore_index=True)
    df = TSDataset.to_dataset(df)

    timestamp = pd.date_range("2020-12-01", "2021-02-11")

    df_1 = pd.DataFrame(
        {
            "timestamp": timestamp,
            "regressor_1": np.arange(len(timestamp)),
            "regressor_2": np.zeros(len(timestamp)),
            "regressor_3": rng.normal(loc=0, scale=1.0, size=len(timestamp)),
            "regressor_4": rng.uniform(low=0, high=10.0, size=len(timestamp)),
            "regressor_5": rng.exponential(scale=1.0, size=len(timestamp)),
            "cat_feature": "hello",
            "segment": "1",
        }
    )
    df_2 = pd.DataFrame(
        {
            "timestamp": timestamp[5:],
            "regressor_1": np.sin(-np.arange(len(timestamp) - 5)),
            "regressor_2": np.log(np.arange(1, len(timestamp) - 4)),
            "regressor_3": rng.normal(loc=0, scale=2.0, size=len(timestamp) - 5),
            "regressor_4": rng.uniform(low=0, high=5.0, size=len(timestamp) - 5),
            "regressor_5": rng.exponential(scale=0.1, size=len(timestamp) - 5),
            "cat_feature": "bye",
            "segment": "2",
        }
    )
    df_exog = pd.concat([df_1, df_2], ignore_index=True)
    df_exog = TSDataset.to_dataset(df_exog)

    return df, df_exog


def main():
    df, df_exog = simple_df_relevance()
    ts = TSDataset(df=df, df_exog=df_exog, known_future="all", freq="D")
    statistics_relevance_table = StatisticsRelevanceTable()
    model_relevance = RandomForestRegressor(n_estimators=100, random_state=42)
    model_relevance_table = ModelRelevanceTable()

    plot_feature_relevance(
        ts=ts,
        relevance_table=statistics_relevance_table,
        normalized=False,
        relevance_aggregation_mode="per-segment",
        top_k=None,
        segments=None,
        columns_num=2,
        figsize=(10, 5),
    )

    plt.savefig("per-segment.png")

    plot_feature_relevance(
        ts=ts,
        relevance_table=statistics_relevance_table,
        normalized=False,
        relevance_aggregation_mode="mean",
        top_k=None,
        segments=None,
        columns_num=2,
        figsize=(10, 5),
    )

    plt.savefig("mean.png")

    transforms = [
        FilterFeaturesTransform(exclude=["cat_feature"]),
        DateFlagsTransform(day_number_in_week=True, day_number_in_month=True, is_weekend=True, out_column="date_flag"),
    ]
    ts.fit_transform(transforms)

    plot_feature_relevance(
        ts=ts,
        relevance_table=model_relevance_table,
        normalized=False,
        relevance_aggregation_mode="per-segment",
        relevance_params={"model": model_relevance},
        top_k=5,
        segments=None,
        columns_num=2,
        figsize=(10, 5),
    )

    plt.savefig("random_forest_per_segment.png")

    plot_feature_relevance(
        ts=ts,
        relevance_table=model_relevance_table,
        normalized=False,
        relevance_aggregation_mode="mean",
        relevance_params={"model": model_relevance},
        top_k=5,
        segments=None,
        columns_num=2,
        figsize=(10, 5),
    )

    plt.savefig("random_forest_mean.png")

    plot_feature_relevance(
        ts=ts,
        relevance_table=model_relevance_table,
        normalized=True,
        relevance_aggregation_mode="mean",
        relevance_params={"model": model_relevance},
        top_k=5,
        segments=None,
        columns_num=2,
        figsize=(10, 5),
    )

    plt.savefig("random_forest_mean_normalized.png")


if __name__ == "__main__":
    main()
```

Статистическая таблица, по сегментам:
per-segment

Статистическая таблица, усреднение: 
 mean

Таблица по модели, по сегментам:
random_forest_per_segment

Таблица по сегментам, усреднение: 
random_forest_mean

Таблица по сегментам, усреднение с нормализацией:
random_forest_mean_normalized

**Review by alex-hse-repository — APPROVED** [review]
